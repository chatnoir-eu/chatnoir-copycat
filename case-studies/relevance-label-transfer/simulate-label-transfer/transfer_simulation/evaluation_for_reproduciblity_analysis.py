import argparse
import glob
import json
from trectools import TrecRun, TrecQrel, TrecEval

QREL_DIR = '../src/main/resources/artificial-qrels/'

RUN_FILE_DIR_TO_QRELS = {
    '/topics-1-50-cw09': QREL_DIR + 'qrels.inofficial.duplicate-free.web.1-50.txt',
    '/topics-1-50-cw12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12.1-50.txt',
    '/topics-1-50-cw12wb12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12wb12.1-50.txt',

    '/topics-51-100-cw09': QREL_DIR + 'qrels.inofficial.duplicate-free.web.51-100.txt',
    '/topics-51-100-cw12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12.51-100.txt',
    '/topics-51-100-cw12wb12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12wb12.51-100.txt',

    '/topics-101-150-cw09': QREL_DIR + 'qrels.inofficial.duplicate-free.web.101-150.txt',
    '/topics-101-150-cw12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12.101-150.txt',
    '/topics-101-150-cw12wb12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12wb12.101-150.txt',

    '/topics-151-200-cw09': QREL_DIR + 'qrels.inofficial.duplicate-free.web.151-200.txt',
    '/topics-151-200-cw12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12.151-200.txt',
    '/topics-151-200-cw12wb12': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cw12wb12.151-200.txt',

    '/topics-201-250-cw12': QREL_DIR + 'qrels.inofficial.duplicate-free.web.201-250.txt',
    '/topics-251-300-cw12': QREL_DIR + 'qrels.inofficial.duplicate-free.web.251-300.txt',

    '/topics-1-50-cc15': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cc15.1-50.txt',
    '/topics-51-100-cc15': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cc15.51-100.txt',
    '/topics-101-150-cc15': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cc15.101-150.txt',
    '/topics-151-200-cc15': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cc15.151-200.txt',
    '/topics-201-250-cc15': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cc15.201-250.txt',
    '/topics-251-300-cc15': QREL_DIR + 'qrels.inofficial.duplicate-free.transferred-to-cc15.251-300.txt',
}


def parse_args():
    parser = argparse.ArgumentParser(description="TBD :)")
    parser.add_argument("--inputDir", type=str, required=True)
    parser.add_argument("--outputFile", type=str, required=True)

    return parser.parse_args()


def extract_corpus(run_file_name):
    return (run_file_name.split('/final-rankings'))[0].split('-')[-1]


def extract_topics(run_file_name):
    return 'topics-' + (run_file_name.split('-' + extract_corpus(run_file_name)))[0].split('topics-')[-1]


def report_run(qrels, run_file_name, remove_docs_with_zero_score=False):
    run = TrecRun(run_file_name)
    system = run.run_data['system'][0]
    if remove_docs_with_zero_score:
        run.run_data = run.run_data[run.run_data['score'] > 0]

    trec_eval = TrecEval(run, qrels)

    ret = {
        'corpus': extract_corpus(run_file_name),
        'topics': extract_topics(run_file_name),
        'tag': system,
        "bpref": trec_eval.getBpref(),
        "pseudoNDCG@10": trec_eval.getNDCG(depth=10),
        "pseudoNDCG": trec_eval.getNDCG()
    }

    return json.dumps(ret)


def report_run_per_query(qrels, run_file_name, remove_docs_with_zero_score=False):
    run = TrecRun(run_file_name)
    system = run.run_data['system'][0]
    if remove_docs_with_zero_score:
        run.run_data = run.run_data[run.run_data['score'] > 0]

    trec_eval = TrecEval(run, qrels)

    bpref = trec_eval.getBpref(per_query=True)
    ndcg_10 = trec_eval.getNDCG(depth=10, per_query='query')
    ndcg = trec_eval.getNDCG(per_query='query')

    ret = bpref.join(ndcg_10, on='query')
    ret = ret.join(ndcg, on='query')

    for query, r in ret.iterrows():
        yield json.dumps({
            'corpus': extract_corpus(run_file_name),
            'topic': query,
            'tag': system,
            "bpref": r['Bpref@1000'],
            "pseudoNDCG@10":  r['NDCG@10'],
            "pseudoNDCG":  r['NDCG@1000']
        })


if __name__ == '__main__':
    args = parse_args()
    with open(args.outputFile + '-per-query-zero-scores-removed.jsonl', 'w+') as f:
        for run_file_dir in RUN_FILE_DIR_TO_QRELS:
            trec_qrel = TrecQrel(RUN_FILE_DIR_TO_QRELS[run_file_dir])
            for run_file in glob.glob(args.inputDir + run_file_dir + '/final-rankings/*.txt', recursive=True):
                print('Evaluate ' + run_file)
                for report_line in report_run_per_query(trec_qrel, run_file, remove_docs_with_zero_score=True):
                    f.write(report_line + '\n')

    with open(args.outputFile + '-zero-scores-removed.jsonl', 'w+') as f:
        for run_file_dir in RUN_FILE_DIR_TO_QRELS:
            trec_qrel = TrecQrel(RUN_FILE_DIR_TO_QRELS[run_file_dir])
            for run_file in glob.glob(args.inputDir + run_file_dir + '/final-rankings/*.txt', recursive=True):
                print('Evaluate ' + run_file)
                f.write(report_run(trec_qrel, run_file, remove_docs_with_zero_score=True) + '\n')

    with open(args.outputFile + '-per-query.jsonl', 'w+') as f:
        for run_file_dir in RUN_FILE_DIR_TO_QRELS:
            trec_qrel = TrecQrel(RUN_FILE_DIR_TO_QRELS[run_file_dir])
            for run_file in glob.glob(args.inputDir + run_file_dir + '/final-rankings/*.txt', recursive=True):
                print('Evaluate ' + run_file)
                for report_line in report_run_per_query(trec_qrel, run_file):
                    f.write(report_line + '\n')

    with open(args.outputFile + '.jsonl', 'w+') as f:
        for run_file_dir in RUN_FILE_DIR_TO_QRELS:
            trec_qrel = TrecQrel(RUN_FILE_DIR_TO_QRELS[run_file_dir])
            for run_file in glob.glob(args.inputDir + run_file_dir + '/final-rankings/*.txt', recursive=True):
                print('Evaluate ' + run_file)
                f.write(report_run(trec_qrel, run_file) + '\n')

