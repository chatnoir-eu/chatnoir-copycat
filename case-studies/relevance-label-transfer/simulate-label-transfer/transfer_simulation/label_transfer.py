import pandas as pd
from trectools import TrecQrel


class BaseLabelTransfer:
    def keep_doc(self, doc_id):
        return doc_id in self.ids_to_transfer


class CW12UrlLabelTransfer(BaseLabelTransfer):
    def __init__(self, input_files):
        df = load_input_files_to_dataframe(input_files)
        df = df[(df['cw12Matches'] > 0)]
        self.ids_to_transfer = df_to_ids_to_transfer(df)
        self.name = 'cw12-url'


class CW12WaybackUrlLabelTransfer(BaseLabelTransfer):
    def __init__(self, input_files):
        df = load_input_files_to_dataframe(input_files)
        df = df[(df['matches'] > 0) | (df['non200Matches'] > 0) | (df['cw12Matches'] > 0)]
        self.ids_to_transfer = df_to_ids_to_transfer(df)
        self.name = 'wayback-cw12-url'


def df_to_ids_to_transfer(df):
    return {str(i) for i in df['document']}


def load_input_files_to_dataframe(input_files):
    dfs = []
    for input_file in input_files:
        dfs += [pd.read_json(input_file, lines=True)]
    return pd.concat(dfs)


def qrels_as_str(input_file, labels_to_keep=None):
    return parse_qrels(input_file, labels_to_keep).qrels_data[["query", "q0", "docid", "rel"]].to_csv(
        sep=" ", header=False, index=False)


def parse_qrels(input_file, labels_to_keep=None):
    ret = TrecQrel(input_file)

    if labels_to_keep is None:
        return ret
    else:
        ret.filename = None

        ret.qrels_data['tmp_delete_me'] = ret.qrels_data['docid'].apply(lambda i: labels_to_keep.keep_doc(i))
        ret.qrels_data = ret.qrels_data[ret.qrels_data['tmp_delete_me']]
        ret.qrels_data = ret.qrels_data.drop(['tmp_delete_me'], axis=1)

        return ret


if __name__ == '__main__':
    files = [
        '/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-05-10-2020/web-2009.jsonl',
        '/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-05-10-2020/web-2010.jsonl',
        '/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-05-10-2020/web-2011.jsonl',
        '/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-05-10-2020/web-2012.jsonl'
    ]
    label_transfers = [
        CW12UrlLabelTransfer(files),
        CW12WaybackUrlLabelTransfer(files)
    ]
    base_dir = 'data/'

    for trec_run_file in ['qrels-web-2009', 'qrels-web-2010', 'qrels-web-2011', 'qrels-web-2012']:
        for label_transfer in label_transfers:
            run_file = base_dir + trec_run_file
            print('Process ' + str(run_file))

            tmp = qrels_as_str(run_file, label_transfer)

            with open(run_file + '-' + label_transfer.name, 'w+') as f:
                f.write(tmp)
