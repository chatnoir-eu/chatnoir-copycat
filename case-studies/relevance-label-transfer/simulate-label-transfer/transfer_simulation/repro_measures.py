def effect_ratio(df, advanced, baseline):

    topics_on_source = set([i for i in df[df['collection'] == 'source'].topic.unique()])
    topics_on_target = set([i for i in df[df['collection'] != 'source'].topic.unique()])
    topics = topics_on_source.intersection(topics_on_target)

    sum_numerator = 0
    sum_denominator = 0

    for topic in topics:
        sum_numerator += improvements_in_experiments(
            df=df,
            advanced=advanced,
            baseline=baseline,
            topic=topic,
            on_old_collection=False
        )
        sum_denominator += improvements_in_experiments(
            df=df,
            advanced=advanced,
            baseline=baseline,
            topic=topic,
            on_old_collection=True
        )

    return (sum_numerator/len(topics))/(sum_denominator/len(topics))


def improvements_in_experiments(df, advanced, baseline, topic, on_old_collection):
    if on_old_collection:
        df = df[df['collection'] == 'source']
    else:
        df = df[df['collection'] != 'source']

    advanced_score = unique_score_or_fail(df, advanced, topic)
    baseline_score = unique_score_or_fail(df, baseline, topic)

    return advanced_score - baseline_score


def unique_score_or_fail(df, method, topic):
    df = df[(df['topic'].astype(int) == int(topic)) & (df['tag'] == method)]
    if len(df) != 1:
        raise ValueError('Couldnt handle df of length ' + str(len(df)) + ' using '
                         + str(method) + ' under ' + str(topic))

    return df.iloc[0]['measure']


def delta_relative_improvement(df, advanced, baseline):
    return relative_improvement(df, advanced, baseline, True) - relative_improvement(df, advanced, baseline, False)


def classify_repro_pairs(df):
    from collections import Counter
    ret = Counter()

    for _, r in df.iterrows():
        c = Counter(classify_repro_pair(r))
        ret = ret + c

    return dict(ret)


def classify_repro_pair(repro_pair):
    er = repro_pair['effectRatio']
    delta_ri = repro_pair['delta_relative_improvement']

    if er > 0 and delta_ri > 0:
        return {
            'effect-size-success-absolute-scores-failure': 1,
            'effect-size-success-absolute-scores-success': 0,
            'effect-size-failure-absolute-scores-failure': 0,
            'effect-size-failure-absolute-scores-success': 0,
        }
    elif er < 0 and delta_ri < 0:
        return {
            'effect-size-success-absolute-scores-failure': 0,
            'effect-size-success-absolute-scores-success': 0,
            'effect-size-failure-absolute-scores-failure': 0,
            'effect-size-failure-absolute-scores-success': 1,
        }
    elif er > 0 and delta_ri < 0:
        return {
            'effect-size-success-absolute-scores-failure': 0,
            'effect-size-success-absolute-scores-success': 1,
            'effect-size-failure-absolute-scores-failure': 0,
            'effect-size-failure-absolute-scores-success': 0,
        }

    return {
        'effect-size-success-absolute-scores-failure': 0,
        'effect-size-success-absolute-scores-success': 0,
        'effect-size-failure-absolute-scores-failure': 1,
        'effect-size-failure-absolute-scores-success': 0,
    }


def relative_improvement(df, advanced, baseline, on_old_collection):
    a_avg = avg_score(df, advanced, on_old_collection)
    b_avg = avg_score(df, baseline, on_old_collection)

    return (a_avg-b_avg)/b_avg


def avg_score(df, method, on_old_collection):
    if on_old_collection:
        df = df[df['collection'] == 'source']
    else:
        df = df[df['collection'] != 'source']

    df = df[df['tag'] == method]

    return float(df.measure.mean())
