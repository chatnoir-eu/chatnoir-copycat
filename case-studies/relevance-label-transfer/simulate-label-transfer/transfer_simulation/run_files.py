import glob
from trectools import TrecRun, TrecQrel
import pandas as pd
import numpy as np


def list_run_files_of_qrel_file(qrel_file_name):
    run_file_dir = run_files_of_qrel_file(qrel_file_name)

    return glob.glob(run_file_dir + '*.gz')


def run_files_of_qrel_file(qrel_file_name):
    if qrel_file_name.startswith('qrels-web-2009') or '.web.1-50' in qrel_file_name:
        return '/mnt/ceph/storage/data-in-progress/trec-system-runs/trec18/web.adhoc/'
    if qrel_file_name.startswith('qrels-web-2010') or '.web.51-100' in qrel_file_name:
        return '/mnt/ceph/storage/data-in-progress/trec-system-runs/trec19/web.adhoc/'
    if qrel_file_name.startswith('qrels-web-2011') or '.web.101-150' in qrel_file_name:
        return '/mnt/ceph/storage/data-in-progress/trec-system-runs/trec20/web.adhoc/'
    if qrel_file_name.startswith('qrels-web-2012') or '.web.151-200' in qrel_file_name:
        return '/mnt/ceph/storage/data-in-progress/trec-system-runs/trec21/web.adhoc/'
    if '.web.201-250' in qrel_file_name:
        return '/mnt/ceph/storage/data-in-progress/trec-system-runs/trec22/web.adhoc/'
    if '.web.251-300' in qrel_file_name:
        return '/mnt/ceph/storage/data-in-progress/trec-system-runs/trec23/web.adhoc/'

    raise ValueError('Could not handle: ' + str(qrel_file_name))


if __name__ == '__main__':
    label_transfers = ['', 'cw12-url', 'wayback-cw12-url']

    for track in ['qrels-web-2009', 'qrels-web-2010', 'qrels-web-2011', 'qrels-web-2012']:
        track_display_name = track.replace('qrels-', '')
        qrels = {
            'orig': TrecQrel('data/' + track),
            'cw12-url': TrecQrel('data/' + track + '-cw12-url'),
            'wayback-cw12-url': TrecQrel('data/' + track + '-wayback-cw12-url'),
        }

        track_eval_data = []

        for run_file in list_run_files_of_qrel_file(track):
            run = TrecRun(run_file)

            run_file_eval = {
                'run': run.get_runid(),
                'track': track_display_name
            }

            print('Process: ' + track_display_name + '(' + run.get_runid() + ')')

            for qrel in qrels.keys():
                run_file_eval[qrel + '-mean-qrel-coverage-10'] = run.get_mean_coverage(qrels[qrel], 10)
                run_file_eval[qrel + '-mean-qrel-coverage-100'] = run.get_mean_coverage(qrels[qrel], 100)
                run_file_eval[qrel + '-mean-qrel-coverage-1000'] = run.get_mean_coverage(qrels[qrel], 1000)

            track_eval_data += [run_file_eval]

        df_name = '/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-05-10-2020/evaluations/' \
                  + track_display_name + '-fairness.jsonl'
        pd.DataFrame(track_eval_data).to_json(df_name, lines=True, orient='records')


def identify_judgments_to_remove_for_leave_one_out(trec_run_files):
    df = pd.concat([i.run_data for i in trec_run_files])
    df['key'] = np.str(df['query']) + df['docid']
    df = df.sort_values('rank', ascending=True).groupby('key', as_index=False).first()
    df = df[['query', 'rank', 'system', 'docid']]

    return df
