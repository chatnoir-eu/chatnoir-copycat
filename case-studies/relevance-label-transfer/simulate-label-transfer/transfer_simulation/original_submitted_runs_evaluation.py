import json
import argparse
from run_files import list_run_files_of_qrel_file
from trectools import TrecRun, TrecQrel, TrecEval
from evaluation_for_reproduciblity_analysis import QREL_DIR


def parse_args():
    parser = argparse.ArgumentParser(description="TBD :)")
    parser.add_argument("--outputFile", type=str, required=True)

    return parser.parse_args()


def report_run(qrels, corpus, topics, run_file_name):
    run = TrecRun(run_file_name)
    trec_eval = TrecEval(run, qrels)

    ret = {
        'corpus': corpus,
        'topics': topics,
        'tag': run.run_data['system'][0],
        "bpref": trec_eval.getBpref(),
        "pseudoNDCG@10": trec_eval.getNDCG(depth=10, removeUnjudged=True),
        "pseudoNDCG": trec_eval.getNDCG(removeUnjudged=True),
    }

    return json.dumps(ret)


def qrel_files(topics):
    transferred_prefix = QREL_DIR + '/qrels-with-only-transferred-original-labels/qrels.inofficial.duplicate-free.'

    ret = [
        (QREL_DIR + 'qrels.inofficial.duplicate-free.web.' + topics + '.txt', 'cw09'),
        (transferred_prefix + 'transferred-to-cc15.' + topics + '.txt', 'cc15')
    ]

    if topics in ['1-50', '51-100', '101-150', '151-200']:
        ret += [
            (transferred_prefix + 'transferred-to-cw12.' + topics + '.txt', 'cw12'),
            (transferred_prefix + 'transferred-to-cw12wb12.' + topics + '.txt', 'cw12wb12')
        ]

    return ret


if __name__ == '__main__':
    args = parse_args()
    with open(args.outputFile, 'w+') as f:
        for topics in ['1-50', '51-100', '101-150', '151-200', '201-250', '251-300']:
            qrel_files_for_topic = qrel_files(topics)
            for qrel_file, corpus in qrel_files_for_topic:
                trec_qrel = TrecQrel(qrel_file)
                for run_file in list_run_files_of_qrel_file(qrel_files_for_topic[0][0]):
                    print(corpus + ': evaluate ' + run_file)
                    f.write(report_run(trec_qrel, corpus, topics, run_file) + '\n')
