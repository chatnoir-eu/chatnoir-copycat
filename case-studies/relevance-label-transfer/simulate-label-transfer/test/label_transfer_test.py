import unittest

from os import path
from transfer_simulation.label_transfer import CW12UrlLabelTransfer, CW12WaybackUrlLabelTransfer, qrels_as_str
from approvaltests.approvals import verify, verify_as_json


class LabelTransferTest(unittest.TestCase):

    def test_that_ceph_is_mounted(self):
        self.assertTrue(
            path.exists('/mnt/ceph/storage/data-in-progress/kibi9872/sigir2021/data-05-10-2020/'),
            'Ceph must be mounted.'
        )

    @classmethod
    def test_cw12url_for_2009_sample(cls):
        input_files = ['test/resources/small-web-2009-sample.jsonl']
        actual = sorted(CW12UrlLabelTransfer(input_files).ids_to_transfer)

        verify_as_json(actual)

    @classmethod
    def test_cw12url_for_2012_sample(cls):
        input_files = ['test/resources/small-web-2012-sample.jsonl']
        actual = sorted(CW12UrlLabelTransfer(input_files).ids_to_transfer)

        verify_as_json(actual)

    @classmethod
    def test_cw12waybackurl_for_2009_sample(cls):
        input_files = ['test/resources/small-web-2009-sample.jsonl']
        actual = sorted(CW12WaybackUrlLabelTransfer(input_files).ids_to_transfer)

        verify_as_json(actual)

    @classmethod
    def test_cw12waybackurl_for_2009_and_2012_sample(cls):
        input_files = ['test/resources/small-web-2009-sample.jsonl', 'test/resources/small-web-2012-sample.jsonl']
        actual = sorted(CW12WaybackUrlLabelTransfer(input_files).ids_to_transfer)

        verify_as_json(actual)

    def test_name_of_cw12waybackurl(self):
        expected = 'wayback-cw12-url'
        input_files = ['test/resources/small-web-2009-sample.jsonl']
        actual = CW12WaybackUrlLabelTransfer(input_files).name

        self.assertEquals(expected, actual)

    def test_name_of_cw12url(self):
        expected = 'cw12-url'
        input_files = ['test/resources/small-web-2009-sample.jsonl']
        actual = CW12UrlLabelTransfer(input_files).name

        self.assertEquals(expected, actual)

    @classmethod
    def test_parsing_of_qrel(cls):
        actual = qrels_as_str('test/resources/sample-qrel-file.txt')

        verify(actual)

    @classmethod
    def test_parsing_of_qrels_with_filtering(cls):
        input_files = ['test/resources/artificial-sample-for-qrel-filtering.jsonl']
        labels_to_keep = CW12WaybackUrlLabelTransfer(input_files)
        actual = qrels_as_str('test/resources/sample-qrel-file.txt', labels_to_keep)

        verify(actual)

