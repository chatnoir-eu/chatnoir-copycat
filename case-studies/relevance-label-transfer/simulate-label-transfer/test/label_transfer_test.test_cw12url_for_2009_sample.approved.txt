[
    "clueweb09-en0006-84-32970",
    "clueweb09-enwp03-03-17640"
]
